scheduler/scheduler.go

```diff
@@ -220,10 +225,154 @@ func spawnHosts(ctx context.Context, newHostsNeeded map[string]int) (map[string]
+			return cloud.HostOptions{}, err
+		}
+		hostOptions := cloud.HostOptions{
+			ParentID: parent.Id,
+			UserName: evergreen.User,
+			UserHost: false,
+		}
+		return hostOptions, nil
+	}
+
+	hostOptions := cloud.HostOptions{
+		UserName: evergreen.User,
+		UserHost: false,
+	}
+	return hostOptions, nil
+}
+
+// generateParentHostOptions generates host options for a parent host
+func generateParentHostOptions() cloud.HostOptions {
+	return cloud.HostOptions{
+		HasContainers: true,
+		UserName:      evergreen.User,
+		UserHost:      false,
+	}
+}
+
+// FindAvailableParent finds a parent host that can accommodate container
+// based on oldest parent
+func findAvailableParent(d distro.Distro) (host.Host, error) {
+	allParents, err := host.FindAllRunningParents()
+	if err != nil {
+		return host.Host{}, err
+	}
+
+	oldestParent := host.Host{}
+	oldestTime := time.Now().Add(time.Hour * 1)
+	for _, parent := range allParents {
+		currentContainers, err := parent.GetContainers()
+		if err != nil {
+			return host.Host{}, err
+		}
+		if parent.CreationTime.Before(oldestTime) && len(currentContainers) < d.MaxContainers {
+			oldestParent = parent
+			oldestTime = parent.CreationTime
+		}
+	}
+	if oldestParent.Id == "" {
+		return host.Host{}, errors.New("no available parent found for container")
+	}
+	return oldestParent, nil
+
+}
+
+// CalcNewParentsNeeded returns the number of additional parents needed to
+// accommodate new containers
+func calcNewParentsNeeded(numCurrentParents, numContainersNeeded, numExistingContainers int, d distro.Distro) int {
+	if numCurrentParents*d.MaxContainers <= numExistingContainers+numContainersNeeded {
+		return int(math.Ceil(float64(numContainersNeeded) / float64(d.MaxContainers)))
+	}
+	return 0
+}
+
+// parentCapacity calculates number of new parents and new containers to create
+// checks to make sure we do not create more parents than allowed
+func parentCapacity(d distro.Distro, numNewParents, numCurrentParents, numCurrentContainers, numContainersToSpawn int) (int, int) {
+	// if there are already maximum numbers of parents running, only spawn
+	// enough containers to reach maximum capacity
+	if numCurrentParents >= d.PoolSize {
+		numContainersToSpawn = (numCurrentParents * d.MaxContainers) - numCurrentContainers
+		numNewParents = 0
+	}
+	// if adding all new parents results in more parents than allowed, only add
+	// enough parents to fill to capacity
+	if numNewParents+numCurrentParents > d.PoolSize {
+		numNewParents = d.MaxContainers - numCurrentParents
+		// since we are adding parents, do not add any containers on this run
+		numContainersToSpawn = 0
+	}
+	return numNewParents, numContainersToSpawn
+}
+
+// containerCapacity calculates how many containers to make
+// if there are containers to spawn, do not spawn parents in the same run
+// instead spawn as many containers as possible now
+func containerCapacity(d distro.Distro, numCurrentParents, numCurrentContainers, numContainersToSpawn int) int {
```

Comment:
I think this function should _just_ determine how many containers to spawn. Its caller can determine whether to actually spawn them, based on whether `parentCapacity` returns > 0.

Reply:
The function does not actually spawn them, if that's what you mean. The comment is misleading, but it's just checking to make sure we do not create more containers than we can currently fit.